Add missing range conflict check between file ingestion and RefitLevel()

hx235 opened the pull request:
**Context:**
File ingestion never checks whether the key range it acts on overlaps with an ongoing RefitLevel() (used in `CompactRange()` with `change_level=true`). That's because RefitLevel() doesn't register and make its key range known to file ingestion. Though it checks overlapping with other compactions by https://github.com/facebook/rocksdb/blob/7.8.fb/db/external_sst_file_ingestion_job.cc#L998.

RefitLevel() (used in `CompactRange()` with `change_level=true`) doesn't check whether the key range it acts on overlaps with an ongoing file ingestion. That's because file ingestion does not register and make its key range known to other compactions.
- Note that non-refitlevel-compaction (e.g, manual compaction w/o RefitLevel() or general compaction) also does not check key range overlap with ongoing file ingestion for the same reason. 
- But it's fine. Credited to @cbi42's discovery, `WaitForIngestFile` was called by background and foreground compactions. They were introduced in https://github.com/facebook/rocksdb/commit/0f88160f67d36ea30e3aca3a3cef924c3a009be6, https://github.com/facebook/rocksdb/commit/5c64fb67d2fc198f1a73ff3ae543749a6a41f513 and https://github.com/facebook/rocksdb/commit/87dfc1d23e0e16ff73e15f63c6fa0fb3b3fc8c8c.
- Regardless, this PR registers file ingestion like a compaction is a general approach that will also add range conflict check between file ingestion and non-refitlevel-compaction, though it has not been the issue motivated this PR.

Above are bugs resulting in two bad consequences:
- If file ingestion and RefitLevel() creates files in the same level, then range-overlapped files will be created at that level and caught as corruption by `force_consistency_checks=true`
- If file ingestion and RefitLevel() creates file in different levels, then with one further compaction on the ingested file, it can result in two same keys both with seqno 0 in two different levels. Then with iterator's [optimization](https://github.com/facebook/rocksdb/blame/c62f3221698fd273b673d4f7e54eabb8329a4369/db/db_iter.cc#L342-L343) that assumes no two same keys both with seqno 0, it will either break this assertion in debug build or, even worst, return value of this same key for the key after it, which is the wrong value to return, in release build.

Therefore we decide to introduce range conflict check for file ingestion and RefitLevel() inspired from the existing range conflict check among compactions.

**Summary:**
- Treat file ingestion job and RefitLevel() as `Compaction` of new compaction reasons: `CompactionReason::kExternalSstIngestion` and `CompactionReason::kRefitLevel` and register/unregister them.  File ingestion is treated as compaction from L0 to different levels and RefitLevel() as compaction from source level to target level.
- Check for `RangeOverlapWithCompaction` with other ongoing compactions, `RegisterCompaction()` on this "compaction" before changing the LSM state in `VersionStorageInfo`, and `UnregisterCompaction()` after changing. 
- Replace scattered fixes (https://github.com/facebook/rocksdb/commit/0f88160f67d36ea30e3aca3a3cef924c3a009be6, https://github.com/facebook/rocksdb/commit/5c64fb67d2fc198f1a73ff3ae543749a6a41f513 and https://github.com/facebook/rocksdb/commit/87dfc1d23e0e16ff73e15f63c6fa0fb3b3fc8c8c.) that prevents overlapping between file ingestion and non-refit-level compaction with this fix cuz those practices are easy to overlook.
- Misc: logic cleanup, see PR comments 

**Test:**
- New unit test `DBCompactionTestWithOngoingFileIngestionParam*` that failed pre-fix and passed afterwards.
- Made compatible with existing tests, see PR comments
- make check
- [Ongoing] Stress test rehearsal with normal value and aggressive CI value https://github.com/facebook/rocksdb/pull/10761

hx235 commented on db/compaction/compaction.cc:
```diff
@@ -238,7 +238,11 @@ Compaction::Compaction(
+          (_compaction_reason == CompactionReason::kExternalSstIngestion ||
+           _compaction_reason == CompactionReason::kRefitLevel)
+              ? false
+              : IsBottommostLevel(output_level_, vstorage, inputs_)),
```
[Note] Logic in `IsBottommostLevel()` does not consider cases like RefitLevel() and file ingestion. So I explicitly excluded them to avoid touching its logic.

hx235 replied:
The original logic is correct to keep - it's just that the bottommost concept is not applied to RefitLevel (where we can output to L0 when L0 is empty) nor file ingestion (where the input files do not exists in LSM prior to the "file-ingestion compaction). I also  didn't want to relax the logic in `IsBottommostLevel` for `RefitLevel` or file ingestion cuz they are really here for conflict checking, not for anything else. Updated `CompactionReason::kExternalSstIngestion/kRefitLevel` comments.

hx235 commented on db/db_bloom_filter_test.cc:
```diff
@@ -2618,8 +2618,7 @@ TEST_F(DBBloomFilterTest, OptimizeFiltersForHits) {
   compact_options.target_level = 7;
-  ASSERT_TRUE(db_->CompactRange(compact_options, handles_[1], nullptr, nullptr)
-                  .IsNotSupported());
+  ASSERT_OK(db_->CompactRange(compact_options, handles_[1], nullptr, nullptr));
```
[Note] Due to some clean up in RefitLevel, this now returns OK. The result of this CompactRange() does not matter for this UT after inspection.

hx235 replied:
> It's due to the new conditional check where level is 8

Correct.

> It seems that we may incorrectly compute final_output_level

Thanks! This is an accidental finding which I plan to address it in https://github.com/facebook/rocksdb/pull/11041. The test has been using the wrong final_output_level prior my PR anyway.

hx235 commented on db/db_impl/db_impl_compaction_flush.cc:
```diff
@@ -1649,6 +1649,10 @@ Status DBImpl::ReFitLevel(ColumnFamilyData* cfd, int level, int target_level) {
 
   InstrumentedMutexLock guard_lock(&mutex_);
 
+  auto* vstorage = cfd->current()->storage_info();
```
[Note] Decided to early return to avoid `cfd->compaction_picker()->GetRange()` below on empty input.

hx235 commented on db/external_sst_file_basic_test.cc:
```diff
@@ -800,6 +800,7 @@ TEST_P(ExternalSSTFileBasicTest, IngestFileWithMultipleValueType) {
   bool verify_checksums_before_ingest = std::get<1>(GetParam());
   do {
     Options options = CurrentOptions();
+    options.disable_auto_compactions = true;
```
[Note] With this PR, file ingestion results will depend on on-going compaction too. Disable auto ones so this test is remained untouched. Also auto-compaction is irrelevant to this test after inspection.

hx235 commented on db/external_sst_file_ingestion_job.h:
```diff
@@ -196,6 +224,18 @@ class ExternalSstFileIngestionJob {
+  //
+  // FileMetaData used in intputs of compactions equivelant to this ingestion
+  // job
+  std::vector<FileMetaData*> compaction_input_metdatas_;
+  // Map from output level to input of compactions equivelant to this ingestion
+  // job
+  std::map<int, std::vector<CompactionInputFiles>>
+      output_level_to_file_ingesting_compaction_input_;
+  // Compactions equivelant to this ingestion job
+  std::vector<Compaction*> file_ingesting_compactions_;
```
[Note] Wish I could have used std::unique_ptr<> for `compaction_input_metdatas_` and `file_ingesting_compactions_` and avoid explicit deletion in ~ExternalSstFileIngestionJob(). But this mysterious window failure  [] stopped me from doing so.

hx235 commented on db/external_sst_file_test.cc:
```diff
@@ -973,7 +973,7 @@ TEST_F(ExternalSSTFileTest, MultiThreaded) {
   do {
     Options options = CurrentOptions();
-
+    options.disable_auto_compactions = true;
```
[Note] Same as above

hx235 commented on db/external_sst_file_test.cc:
```diff
@@ -1249,8 +1249,9 @@ TEST_P(ExternalSSTFileTest, PickedLevel) {
   // output of compaction going to L1
-  ASSERT_OK(GenerateAndAddExternalFile(options, {4, 7}, -1, false, false, true,
-                                       false, false, &true_data));
+  ASSERT_OK(GenerateAndAddExternalFile(options, {4, 7}, -1,
```
[Note] Similar to above, ingest file of overlapped ranges with ongoing compaction needs `allow_global_seqno=true` just like how it's needed for overlapping ranges with existing files.

hx235 commented:
@cbi42  minor update to test titles and pr title/summary

hx235 commented:
@cbi42  - as mentioned offline, I think we should try to remove existing scattered practice like "WaitForIngestFile" for background and foreground executed compactions if we could. Will try to do that and let you know here when it's done!

cbi42 commented on db/external_sst_file_ingestion_job.h:
```diff
@@ -99,6 +102,21 @@ class ExternalSstFileIngestionJob {
+      cfd_->compaction_picker()->UnregisterCompaction(c);
+      delete c;
+    }
+    file_ingesting_compactions_.clear();
```
I think the `clear()`s here are not needed.

hx235 replied:
Fixed

cbi42 reviewed (commented):
Nice fix! LGTM overall, will look at tests next.

hx235 commented on db/compaction/compaction_picker.cc:
```diff
@@ -1126,7 +1126,11 @@ void CompactionPicker::RegisterCompaction(Compaction* c) {
          !FilesRangeOverlapWithCompaction(*c->inputs(), c->output_level(),
                                           c->GetPenultimateLevel()));
-  if (c->start_level() == 0 ||
+  // File ingestion's start level is just a placeholder number without actual
+  // meaning as file ingestion technically does not have an input level like
+  // other compactions
+  if ((c->start_level() == 0 &&
+       c->compaction_reason() != CompactionReason::kExternalSstIngestion) ||
```
File ingestion is treated as a Compaction with input level L0 as a placeholder number because technically file ingestion does not have an input level and we need a number for populating Compaction.

(I thought of using -1 but its actually the Compaction::kInvalidLevel used for error checking or exclusion. )

As a result of this placeholder number "0", we can't see file ingestion as level 0 compaction otherwise every file ingestion can block other innocent compactions (e.g, https://github.com/facebook/rocksdb/blob/7.8.fb/db/compaction/compaction_picker.cc#L646, https://github.com/facebook/rocksdb/blob/7.8.fb/db/compaction/compaction_picker_level.cc#L716.

hx235 commented on db/db_compaction_test.cc:
```diff
@@ -6181,125 +6181,228 @@ TEST_P(DBCompactionTestWithParam, FixFileIngestionCompactionDeadlock) {
+    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
+  }
 
-TEST_F(DBCompactionTest, IngestFileOfRangeOverlappedWithOngoingRefitLevel) {
```
I realized this UT is not needed. I was planning to use this UT to demonstrate a corruption due to
```
timeline  t1 (refitlevel)                          t2 (file ingestion)
t0   |Decides to refit from Ln+1 to Ln
t1   |Release lock in LogAndApply()
t2   |                                             [a]Mistakenly decides to output to Ln cuz t1 hasn't updated LSM tree
t3   |                                             Release lock
t4   |Acquire lock 
```

However [a] couldn't happen because file ingestion actually LogAndApply() twice, first-time in https://github.com/facebook/rocksdb/blob/7.8.fb/db/db_impl/db_impl.cc#L5858 (before deciding level) and second-time in https://github.com/facebook/rocksdb/blob/7.8.fb/db/db_impl/db_impl.cc#L5352 (after deciding level).

Due to manifest writers are in order, t1's LogAndApply() will finishes before t2's first-time LogAndApply() finishes (hence t2 deciding the output level for file ingestion).Therefore by the time t2 gets to decide the output level, it already has the updated LSM after t1 so overlap-range corruption problem should not occur. 

Another way to see this is to remove my "cfd->compaction_picker()->RegisterCompaction(c.get());" for registering RefitLevel() and run the old test `DBCompactionTest, IngestFileOfRangeOverlappedWithOngoingRefitLevel`.

hx235 commented:
A few update @cbi42 :
1. Removed previous `WaitForIngestFile()` fix to range conflict between file ingestion and CompactFiles/Compaction/CompactRange https://github.com/facebook/rocksdb/commit/0f88160f67d36ea30e3aca3a3cef924c3a009be6, https://github.com/facebook/rocksdb/commit/5c64fb67d2fc198f1a73ff3ae543749a6a41f513 and https://github.com/facebook/rocksdb/commit/87dfc1d23e0e16ff73e15f63c6fa0fb3b3fc8c8c.  Replaced an old test `TEST_F(ExternalSSTFileTest, PickedLevelBug)` written based on the old assumption that we waits for file ingestion with new test `DBCompactionTestWithOngoingFileIngestionParam`
2. A bug fix in https://github.com/facebook/rocksdb/pull/10988/files#diff-c261f77d6dd2134333c4a955c311cf4a196a08d3c2bb6ce24fd6801407877c89R1133
3. Included my `TEST_F(DBCompactionTest, RefitLevelOverlappedWithOngoingFileIngestion)` as part of the DBCompactionTestWithOngoingFileIngestionParam`. Did some cleanup and clarification for readability and being able to see the failure without the fix.
4. Removed my `TEST_F(DBCompactionTest, IngestFileOfRangeOverlappedWithOngoingRefitLevel)` as I realized there will not be corruption due to file ingestion happens when the ongoing refit level releases the lock - see PR comment https://github.com/facebook/rocksdb/pull/10988/commits/fd2ba40f2ac8d94346dd3d9251fedc1f7bb25464#r1051334645

hx235 commented on db/external_sst_file_ingestion_job.cc:
```diff
@@ -795,8 +858,16 @@ Status ExternalSstFileIngestionJob::AssignLevelAndSeqnoForIngestedFile(
     }
-
-    if (vstorage->NumLevelFiles(lvl) > 0) {
+    if (cfd_->RangeOverlapWithCompaction(
```
Done.

hx235 commented on db/external_sst_file_ingestion_job.cc:
```diff
@@ -873,15 +945,24 @@ Status ExternalSstFileIngestionJob::AssignLevelAndSeqnoForIngestedFile(
   int bottom_lvl = cfd_->NumberLevels() - 1;
+  // First, check if new files range-overlap with ongoing compaction's output to
+  // bottommost level
+  if (cfd_->RangeOverlapWithCompaction(
```
This is moved from the `IngestedFileFitInLevel` below.

hx235 replied:
Ok - will remove for better semantics.

hx235 replied:
Fixed

hx235 commented on db/external_sst_file_ingestion_job.h:
```diff
@@ -196,6 +224,18 @@ class ExternalSstFileIngestionJob {
+
+  // Below are variables used in (un)registering range for this ingestion job
+  //
+  // FileMetaData used in intputs of compactions equivelant to this ingestion
```
Fixed - been spelling "equivalent" wrong my whole life so far  ....

hx235 commented on include/rocksdb/options.h:
```diff
@@ -1933,7 +1933,8 @@ struct IngestExternalFileOptions {
   // If set to false, IngestExternalFile() will fail if the file key range
-  // overlaps with existing keys or tombstones in the DB.
+  // overlaps with existing keys or output of ongoing compaction during file
+  // ingestion or tombstones in the DB.
```
Fixed

hx235 commented on db/db_compaction_test.cc:
```diff
@@ -6181,6 +6181,230 @@ TEST_P(DBCompactionTestWithParam, FixFileIngestionCompactionDeadlock) {
+
+  // To coerce following sequence of events
+  // Timeline   Thread 1 (Ingest s2)          Thread 2 (Compact s1)
+  // t0   |     Decide to output to Ln
```
[minor-TODO] Might be more clear to replace Ln with actual level number

hx235 commented on db/db_compaction_test.cc:
```diff
@@ -6181,6 +6181,230 @@ TEST_P(DBCompactionTestWithParam, FixFileIngestionCompactionDeadlock) {
+  }
+
+  if (compaction_path_to_test_ != "RefitLevelCompactRange") {
+    SetupLastLevelToForceFilesIngestionToPenultimateLevel();
```
[minor-TODO] Maybe a less verbose function name - as long as reader knows this is a setup for our test purpose.

hx235 commented on db/db_compaction_test.cc:
```diff
@@ -6181,6 +6181,230 @@ TEST_P(DBCompactionTestWithParam, FixFileIngestionCompactionDeadlock) {
+    ASSERT_OK(db_->IngestExternalFile({s2}, IngestExternalFileOptions()));
+  });
+
+  // Compact s2. Withouth proper range conflict check,
```
[TODO] Compact `s1`

hx235 commented on db/compaction/compaction.cc:
```diff
@@ -259,8 +261,15 @@ Compaction::Compaction(
-      penultimate_level_(EvaluatePenultimateLevel(
-          vstorage, immutable_options_, start_level_, output_level_)) {
+      penultimate_level_(
+          // For simplicity, we won't apply the concept of penultimate_level to
+          // `CompactionReason::kExternalSstIngestion` and
+          // `CompactionReason::kRefitLevel`
+          _compaction_reason == CompactionReason::kExternalSstIngestion ||
+                  _compaction_reason == CompactionReason::kRefitLevel
```
[Note] Excluded these two for the concept of `penultimate_level_` to avoid failure in https://github.com/facebook/rocksdb/blob/7.8.fb/db/compaction/compaction_picker.cc#L1138, where input of refit level and file ingestion overlap with the second to last level, which is fine for refit level and ingestion

hx235 commented on db/db_impl/db_impl_compaction_flush.cc:
```diff
@@ -1972,9 +1969,6 @@ Status DBImpl::RunManualCompaction(
                manual.begin, manual.end, &manual.manual_end, &manual_conflict,
                max_file_num_to_ignore, trim_ts)) == nullptr &&
           manual_conflict))) {
-      // exclusive manual compactions should not see a conflict during
-      // CompactRange
-      assert(!exclusive || !manual_conflict);
```
[Note] Unfortunately, exclusive manual compaction can have *manual_conflict == true since it can conflict with file ingestion after we add  file ingestion to conflict check ADN remove WaitForIngestFile(); I think it's fine to remove the assertion as it's more of a defensive assertion.

hx235 commented on db/compaction/compaction.cc:
```diff
@@ -255,8 +262,15 @@ Compaction::Compaction(
-      penultimate_level_(EvaluatePenultimateLevel(
-          vstorage, immutable_options_, start_level_, output_level_)) {
+      penultimate_level_(
+          // For simplicity, we don't support the concept of "penultimate level"
+          // with `CompactionReason::kExternalSstIngestion` and
+          // `CompactionReason::kRefitLevel`
+          _compaction_reason == CompactionReason::kExternalSstIngestion ||
+                  _compaction_reason == CompactionReason::kRefitLevel
+              ? kInvalidLevel
```
[Note] Excluded these two for the concept of `penultimate_level_` to avoid failure in https://github.com/facebook/rocksdb/blob/7.8.fb/db/compaction/compaction_picker.cc#L1138, where input of refit level and file ingestion overlap with the second to last level, which is fine for refit level and ingestion

hx235 commented on db/compaction/compaction.cc:
```diff
@@ -255,8 +262,15 @@ Compaction::Compaction(
+          // `CompactionReason::kRefitLevel`
+          _compaction_reason == CompactionReason::kExternalSstIngestion ||
+                  _compaction_reason == CompactionReason::kRefitLevel
+              ? Compaction::kInvalidLevel
```
[Note] Excluded these two for the concept of penultimate_level_ to avoid failure in https://github.com/facebook/rocksdb/blob/7.8.fb/db/compaction/compaction_picker.cc#L1138, where input of refit level and file ingestion overlap with the second to last level, which is fine for refit level and ingestion

hx235 commented:
Update: ongoing stress test has revealed a few more places to fix to replace WaitForIngestFiles()

hx235 commented on db/db_impl/db_impl_compaction_flush.cc:
```diff
@@ -1197,6 +1197,9 @@ Status DBImpl::CompactRangeInternal(const CompactRangeOptions& options,
                    "[RefitLevel] waiting for background threads to stop");
+    // TODO(hx235): remove `Enable/DisableManualCompaction` and
+    // `Continue/PauseBackgroundWork` once we have better conflict checking
+    // support for `RefitLevel()`
```
Fixed

hx235 commented on db/compaction/compaction_picker_level.cc:
```diff
@@ -447,21 +447,21 @@ bool LevelCompactionBuilder::SetupOtherInputsIfNeeded() {
       compaction_inputs_.push_back(output_level_inputs_);
     }
 
+    // In some edge cases we could pick a compaction that will be compacting
+    // a key range that overlap with another running compaction, and both
+    // of them have the same output level. This could happen if
+    // (1) we are running a non-exclusive manual compaction
+    // (2) AddFile ingest a new file into the LSM tree
+    // We need to disallow this from happening.
+    if (compaction_picker_->FilesRangeOverlapWithCompaction(
+            compaction_inputs_, output_level_,
+            Compaction::EvaluatePenultimateLevel(
+                vstorage_, ioptions_, start_level_, output_level_))) {
+      // This compaction output could potentially conflict with the output
+      // of a currently running compaction, we cannot run it.
+      return false;
+    }
```
[Note] As now we add file ingestion/refit level into range conflict check, I don't see why we can exclude l0_trivial_move from the check. So I decided to do check regardless of whether it's l0_trivial_move.

hx235 commented on db/compaction/compaction_picker_fifo.cc:
```diff
@@ -177,6 +177,18 @@ Compaction* FIFOCompactionPicker::PickSizeCompaction(
               ,
               max_compact_bytes_per_del_file,
               mutable_cf_options.max_compaction_bytes, &comp_inputs)) {
+        if (FilesRangeOverlapWithCompaction(
+                {comp_inputs}, 0 /* output_level */,
+                Compaction::EvaluatePenultimateLevel(
+                    vstorage, ioptions_, comp_inputs.level /* start_level */,
+                    0 /* output_level */))) {
+          ROCKS_LOG_BUFFER(log_buffer,
+                           "[%s] FIFO compaction: Conflict with an ongoing "
+                           "compaction or file ingestion",
+                           cf_name.c_str());
+          return nullptr;
+        }
```
[Note] Surprisingly, FIFO is not doing deletion compaction `* is deletion compaction */ false`, which means it will output files and can potentially conflict with file ingestion or refit level. Therefore I added the conflict check here. Same for below

cbi42 commented on db/db_compaction_test.cc:
```diff
@@ -6181,6 +6181,231 @@ TEST_P(DBCompactionTestWithParam, FixFileIngestionCompactionDeadlock) {
+      std::string end_key = "k4";
+      Slice end(end_key);
+      TEST_SYNC_POINT("PreCompaction");
+      // Withouth proper range conflict check,
```
nit: without (my IDE is catching these, not me : ))

hx235 replied:
Time to get myself something like that!!!

cbi42 commented on db/db_compaction_test.cc:
```diff
@@ -6181,6 +6181,231 @@ TEST_P(DBCompactionTestWithParam, FixFileIngestionCompactionDeadlock) {
+    } else {
+      options_.num_levels = 3;
+    }
+    options_.compaction_style = CompactionStyle::kCompactionStyleLevel;
```
Tests look great. I assume the issue also applies to other compaction styles?

hx235 replied:
I doubt for other compactions. Assume the WaitForIngestFile() is already removed, in order for the issue happens, we need a compaction to output files to the output level of file ingestion. The compaction could be user-specified-level compaction (RefitLevel, CompactFiles(level)) or non-user-specified-level compaction (all other compactions.)

hx235 replied:
> I assume the issue also applies to other compaction styles?

I doubt it for other compaction styles. 

For multi-level FIFO, the non-deletion compaction (the one has output files i.e,  FindIntraL0Compaction in PickSizeCompaction, PickCompactionToWarm, CompactFiles(output_level <= FIFOCompactionPicker::MaxOutputLevel())) are all L0->L0 compaction. So compaction output will stay in L0 where overlapping range is allowed as well as we won't have inverted-level for new/old data. 

For universal, it's a bit tricky. 
- For RefitLevel(change_level=true), the one our PR is motivated to fix, I believe universal has been safe. This is because https://github.com/facebook/rocksdb/blob/7.8.fb/db/external_sst_file_ingestion_job.cc#L799 leads to files ingested to non-L0 level only if that level is non-empty, under universal compaction (though I didn't spend time in fully understanding the intention). So the compaction output will output files to a non-empty level, which can't be the target_level of RefitLevel(change_level=true), guaranteed by existing codes.
- For non-RefitLevel() compaction, the one our PR replace the fixes to,  I believe it's possible for universal to hit the issues (without WaitForIngestFiles() nor our fix). It's just not easy to write a UT to ingest file to non-L0 for universal compaction due to https://github.com/facebook/rocksdb/blob/7.8.fb/db/external_sst_file_ingestion_job.cc#L828-L829 and I decided to trust the way we are replacing the previous fix is generic enough for universal compaction.

cbi42 reviewed (approved):
LGTM! Left some comments that should not require significant changes.

hx235 commented on db/db_impl/db_impl_compaction_flush.cc:
```diff
@@ -1704,6 +1751,11 @@ Status DBImpl::ReFitLevel(ColumnFamilyData* cfd, int level, int target_level) {
     Status status = versions_->LogAndApply(cfd, mutable_cf_options, &edit,
                                            &mutex_, directories_.GetDbDir());
 
+    if (!(c == nullptr)) {
```
I guessed I wanted to be safe for .get() but will remove.

hx235 replied:
Fixed

hx235 commented on db/external_sst_file_ingestion_job.cc:
```diff
@@ -477,9 +477,73 @@ Status ExternalSstFileIngestionJob::Run() {
     f_metadata.temperature = f.file_temperature;
     edit_.AddFile(f.picked_level, f_metadata);
   }
+
```
I went this way for my first draft and switched to this version due to some corner case. Let me try to find it again. If not, then I will change to compaction per file :)

hx235 replied:
Oh right - UT like ExternalSSTFileBasicTest.AdjacentRangeDeletionTombstones will fail for compaction per-file cuz 
current `RangeOverlapWithCompaction` will see compactions per ingested file with adjacent range deletion tombstones as overlapping. 

For example, if s1={smallest_range_del_key=300, largest_range_del_key=400, output_level=6} and s2={smallest_range_del_key=400, largest_range_del_key=500, output_level=6} are two compactions registered separately in the same ingestion job, they are seen as overlapping cuz `RangeOverlapWithCompaction` just do simple `ucmp->CompareWithoutTimestamp`. 

I recalled there is some logic dealing with case like that but they were deep in internal iterator level and can't be easily called in RangeOverlapWithCompaction.  I also didn't want to make RangeOverlapWithCompaction aware of key type to make an exception. So I decided to do per-level compaction so  s1={smallest_range_del_key=300, largest_range_del_key=400} and s2={smallest_range_del_key=400, largest_range_del_key=500} are not two compactions but one compaction to same level.

Added TODO and a minor refactory.

cbi42 replied:
I see. For checking range deletion overlap, internal key is needed (see `sstableKeyCompare`), so this makes sense.
